test/jdk/sun/security/pkcs11/Cipher/CancelMultipart.java

```diff
@@ -0,0 +1,200 @@
+            if (mode == DECRYPT_MODE) {
+                cipher.update(new byte[1]);
+                cipher.doFinal(new byte[1], 0, 0);
+            } else {
+                cipher.update(new byte[1]);
+            }
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        main(new CancelMultipart(), args);
+    }
+
+    @Override
+    public void main(Provider p) throws Exception {
+        init(p);
+
+        // Try multiple paths:
+
+        executeTest(new SessionLeaker(new LeakEncrypt(), new LeakByteArray()),
+                "P11Cipher::implUpdate(byte[], int, int, byte[], int, int)");
+
+        executeTest(new SessionLeaker(new LeakEncrypt(), new LeakByteBuffer()),
+                "P11Cipher::implUpdate(ByteBuffer, ByteBuffer)");
+
+        executeTest(new SessionLeaker(new LeakDecrypt(), new LeakByteArray()),
+                "P11Cipher::implDoFinal(byte[], int, int)");
+
+        executeTest(new SessionLeaker(new LeakDecrypt(), new LeakByteBuffer()),
+                "P11Cipher::implDoFinal(ByteBuffer)");
+
+        System.out.println("TEST PASS - OK");
+    }
+
+    private static void executeTest(SessionLeaker sl, String testName)
+            throws Exception {
+        try {
+            sl.leakAndTry();
+            System.out.println(testName +  ": OK");
+        } catch (Exception e) {
+            System.out.println(testName +  ": FAILED");
+            throw e;
+        }
+    }
+
+    private static void init(Provider p) throws Exception {
+        provider = p;
+
+        // The max number of sessions is 2 because, in addition to the
+        // operation (i.e. PKCS11::getNativeKeyInfo), a session to hold
+        // the P11Key object is needed.
+        setMaxSessions(2);
+    }
+
+    /*
+     * This method is intended to generate pression on the number of sessions
+     * to be used from the NSS Software Token, so sessions with (potentially)
+     * active operations are reused.
+     */
+    private static void setMaxSessions(int maxSessions) throws Exception {
+        Field tokenField = Class.forName("sun.security.pkcs11.SunPKCS11")
+                .getDeclaredField("token");
+        tokenField.setAccessible(true);
+        Field sessionManagerField = Class.forName("sun.security.pkcs11.Token")
+                .getDeclaredField("sessionManager");
+        sessionManagerField.setAccessible(true);
+        Field maxSessionsField = Class.forName("sun.security.pkcs11.SessionManager")
+                .getDeclaredField("maxSessions");
+        maxSessionsField.setAccessible(true);
+        Object sessionManagerObj = sessionManagerField.get(
+                tokenField.get(provider));
+        maxSessionsField.setInt(sessionManagerObj, maxSessions);
+    }
+
+    private static void tryCipherInit() throws Exception {
+        Cipher cipher = Cipher.getInstance("AES/ECB/NoPadding", provider);
+        cipher.init(Cipher.ENCRYPT_MODE, key);
```

Comment:
Add a comment here or to this method about the CKR_OPERATION_ACTIVE error (w/o this fix) as this is actually the key test point.

Reply:
Right, thanks.